DOCSP-41982: cluster monitoring (#144)
mongodb/docs-php-library: source/monitoring/cluster-monitoring.txt

.. _php-cluster-monitoring:

==================
Cluster Monitoring
==================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, server, topology

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecols

Overview
--------

This guide shows you how to use the {+php-library+} to monitor topology
events in a MongoDB instance, replica set, or sharded cluster. The

> driver creates topology events, also known as Server Discovery and
> Monitoring (**SDAM**) events, when there are any changes in the state of the

Reviewer: S: i know SDAM is the more commonly used term, but i wonder if bolding the whole thing makes more sense

```suggestion
driver creates topology events, also known as **Server Discovery and
Monitoring (SDAM)** events, when there are any changes in the state of the
```

Author: probably ok to go without any bolding, tbh

MongoDB instance or cluster that you are connected to.

You might use information about topology events in your
application to understand cluster changes, assess cluster health, or
perform capacity planning.

Subscribe to Events
-------------------

You can access details about SDAM events by subscribing to them

> in your application. To subscribe to any event, create a class that

Reviewer: S:

```suggestion
in your application. To subscribe to an event, create a class that
```

implements the ``MongoDB\Driver\Monitoring\SDAMSubscriber`` interface,
then use the ``MongoDB\Client::addSubscriber()`` method to register the
event subscriber with your ``MongoDB\Client`` instance.

> The following code creates the ``MySubscriber`` class that implements
> ``SDAMSubscriber`` to output a message when a ``ServerOpeningEvent`` is

Reviewer: I: https://www.mongodb.com/docs/meta/style-guide/writing/use-that-which-correctly/

```suggestion
The following code creates the ``MySubscriber`` class, which implements
``SDAMSubscriber`` to output a message when a ``ServerOpeningEvent`` is
```

Author: reworded

generated by the server:

.. literalinclude:: /includes/monitoring/sdam.php
   :start-after: start-mysubscriber
   :end-before: end-mysubscriber
   :language: php
   :copyable:
   :dedent:

.. note::

   As shown in the preceding code, you must implement all of the methods

>    of ``SDAMSubscriber``, even for events you are not subscribing to.

Reviewer: S:

```suggestion
   of the ``SDAMSubscriber`` interface, even for events you are not subscribing to.
```

   You can implement these methods with empty bodies so that the library
   does not generate any messages for these events.

> Then, the following code uses the ``addSubscriber()`` method to register
> ``MySubscriber`` with the client:

Reviewer: I: I like the use of "then" to signify a subsequent step, but I would tie it to a user step instead of the code example. Something like:

```suggestion
Then, use the ``addSubscriber()`` method to register
``MySubscriber`` with the client, as shown in the following code:
```

.. literalinclude:: /includes/monitoring/sdam.php
   :start-after: start-add-sub
   :end-before: end-add-sub
   :language: php
   :copyable:
   :dedent:

When you run the application, your subscriber records the SDAM event and
outputs messages such as the following:

.. code-block:: none
   :copyable: false

   Server opening on ac-rmuag0v-shard-00-00.gh0qg50.mongodb.net:27017
   Server opening on ac-rmuag0v-shard-00-01.gh0qg50.mongodb.net:27017
   Server opening on ac-rmuag0v-shard-00-02.gh0qg50.mongodb.net:27017

Event Descriptions
------------------

You can subscribe to the following SDAM events by implementing the
corresponding method from the ``SDAMSubscriber`` interface:

.. list-table::
   :widths: 35 65 
   :header-rows: 1

   * - Event Name
     - Description

   * - ``ServerChangedEvent``

>      - Created when an instance state changes, such as from secondary to

Reviewer: Optional: Could remove "Created when" from the start of every event (probably along with a new column header name)

Author: I can't think of a good way to adjust the column header, as renaming it "Situation that Triggers Event" seems worse than keeping the "Created when"s in

Reviewer 2: In the docs for [ServerChangedEvent](https://www.php.net/manual/en/class.mongodb-driver-monitoring-serverchangedevent.php) we say the event class "encapsulates information about a changed server description".

Here, the table is presenting the event as the thing that happened instead of just an object containing some information about a thing that happened.

The "Condition" column in [Events API](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring-logging-and-monitoring.md#events-api-1) from the driver spec might be helpful. Note that the spec includes "Description" in the name of these events, where PHP just calls it ServerChangedEvent. I think that was due to us copying the shorter name from our libmongoc dependency.

I might suggest:

> When the server description changes, such as the server's type changing from secondary to primary.

.. _php-cluster-monitoring:

==================
Cluster Monitoring
==================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, server, topology

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecols

Overview
--------

This guide shows you how to use the {+php-library+} to monitor topology
events in a MongoDB instance, replica set, or sharded cluster. The

> driver creates topology events, also known as Server Discovery and

Reviewer 2: Technically, topology events are those related to the entire cluster, while server events are those pertaining to a single server within the topology. [SDAM monitoring](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring-logging-and-monitoring.md) includes both types of events.

I would personally just refer to them as SDAM events, as you seem to do in a subsequent section. If you need a short name, "cluster events" might also work but there's no prior art for that in any driver spec.

Disregard if this name was intentionally chosen for consistency with existing driver docs.

Author: Good to know, in general it is ok if this page differs from the same page in other docs. Each docs standardization is meant to improve on previous ones!

You might use information about topology events in your
application to understand cluster changes, assess cluster health, or
perform capacity planning.

Subscribe to Events
-------------------

You can access details about SDAM events by subscribing to them
in your application. To subscribe to an event, create a class that
implements the ``MongoDB\Driver\Monitoring\SDAMSubscriber`` interface,
then use the ``MongoDB\Client::addSubscriber()`` method to register the
event subscriber with your ``MongoDB\Client`` instance.

The following code creates the ``MySubscriber`` class, which implements
the ``SDAMSubscriber`` interface. The class is defined with a method to
output a message when a ``ServerOpeningEvent`` is generated by the
server:

.. literalinclude:: /includes/monitoring/sdam.php
   :start-after: start-mysubscriber
   :end-before: end-mysubscriber
   :language: php
   :copyable:
   :dedent:

.. note::

   As shown in the preceding code, you must implement all of the methods
   of the ``SDAMSubscriber`` interface, even for events you are not subscribing to.
   You can implement these methods with empty bodies so that the library

>    does not generate any messages for these events.

Reviewer 2: Technically, the library never "generate(s) any messages". That's something you've chosen to do in the subscriber.

This would be more accurate (feel free to reword):

> This example defines the additional methods as empty so not to output any messages for those events.

Then, use the ``addSubscriber()`` method to register ``MySubscriber``
with the client, as shown in the following code:

.. literalinclude:: /includes/monitoring/sdam.php
   :start-after: start-add-sub
   :end-before: end-add-sub
   :language: php
   :copyable:
   :dedent:

When you run the application, your subscriber records the SDAM event and
outputs messages such as the following:

.. code-block:: none
   :copyable: false

   Server opening on ac-rmuag0v-shard-00-00.gh0qg50.mongodb.net:27017
   Server opening on ac-rmuag0v-shard-00-01.gh0qg50.mongodb.net:27017
   Server opening on ac-rmuag0v-shard-00-02.gh0qg50.mongodb.net:27017

Event Descriptions
------------------

You can subscribe to the following SDAM events by implementing the

> corresponding method from the ``SDAMSubscriber`` interface:

Reviewer 2: This talks about the "corresponding method" from [SDAMSubscriber](https://www.php.net/manual/en/class.mongodb-driver-monitoring-sdamsubscriber.php), but those names aren't presented in the table below. It would be a bit redundant to dedicate a column to the method name, but I think some links to the PHP.net docs could make this a more helpful resource.

Consider working in links to the following:

 * The [SDAMSubscriber](https://www.php.net/manual/en/class.mongodb-driver-monitoring-sdamsubscriber.php) interface.
 * Individual class pages for each event under https://www.php.net/manual/en/mongodb.monitoring.php. For example, ServerOpeningEvent can link to https://www.php.net/manual/en/class.mongodb-driver-monitoring-serveropeningevent.php.

Author: The API doc for SDAMSubscriber is linked at the end of the page, but I will link each event in the table.

.. list-table::
   :widths: 35 65 
   :header-rows: 1

   * - Event Name
     - Description

   * - ``ServerChangedEvent``
     - Created when an instance state changes, such as from secondary to
       primary.

   * - ``ServerOpeningEvent``
     - Created when the server is initialized.

   * - ``ServerClosedEvent``
     - Created when the server is closed.

   * - ``TopologyChangedEvent``
     - Created when the topology changes, such as an election of a new
       primary or disconnection of a ``mongos`` proxy.

   * - ``TopologyOpeningEvent``
     - Created when the topology is initialized.

   * - ``TopologyClosedEvent``
     - Created when the topology is closed.

   * - ``ServerHeartbeatStartedEvent``
     - Created when the heartbeat is started.

   * - ``ServerHeartbeatSucceededEvent``
     - Created when the heartbeat succeeds.

   * - ``ServerHeartbeatFailedEvent``
     - Created when the heartbeat fails.

API Documentation
-----------------

To learn more about any of the classes or methods discussed in this guide, see the
following API documentation:

- :phpmethod:`MongoDB\Client::addSubscriber()`

> - :phpclass:`MongoDB\Client`

Reviewer 2: The `removeSubscriber()` method would be a more helpful reference. I don't see any reason to link to the Client class page directly.

On that note, perhaps you should include a brief example of unregistering a subscriber.

   :copyable: false

   Server opening on ac-rmuag0v-shard-00-00.gh0qg50.mongodb.net:27017
   Server opening on ac-rmuag0v-shard-00-01.gh0qg50.mongodb.net:27017
   Server opening on ac-rmuag0v-shard-00-02.gh0qg50.mongodb.net:27017

Event Descriptions
------------------

You can subscribe to SDAM events by implementing the corresponding
method from the ``SDAMSubscriber`` interface. The following table
provides the name of each SDAM event, linked to the class's API
documentation, and a description of when the event is published:

.. list-table::
   :widths: 35 65 
   :header-rows: 1

   * - Event Type
     - Description

   * - :php:`ServerChangedEvent <mongodb-driver-monitoring-serverchangedevent>`
     - Created when the server description changes, such as the server's type
       changing from secondary to primary.

   * - :php:`ServerOpeningEvent <mongodb-driver-monitoring-serveropeningevent>`
     - Created when a server connection is established.

   * - :php:`ServerClosedEvent <mongodb-driver-monitoring-serverclosedevent>`
     - Created when a server connection is closed.

   * - :php:`TopologyChangedEvent <mongodb-driver-monitoring-topologychangedevent>`

>      - Created when the topology description changes, such when there is

Reviewer 2: I think you accidentally a word in this sentence.

```suggestion
     - Created when the topology description changes, such as when there is
```

       an election of a new primary.

   * - :php:`TopologyOpeningEvent <mongodb-driver-monitoring-topologyopeningevent>`
     - Created when the driver first connects to the cluster.

   * - :php:`TopologyClosedEvent <mongodb-driver-monitoring-topologyclosedevent>`
     - Created when the driver disconnects from the cluster.

   * - :php:`ServerHeartbeatStartedEvent <mongodb-driver-monitoring-serverheartbeatstartedevent>`

>      - Created when the server monitor sends a ``hello`` call to the server.

Reviewer 2: ```suggestion
     - Created when the server monitor sends a ``hello`` command to the server.
```

"Command" is less ambiguous.

Event Descriptions
------------------

You can subscribe to SDAM events by implementing the corresponding
method from the ``SDAMSubscriber`` interface. The following table
provides the name of each SDAM event, linked to the class's API
documentation, and a description of when the event is published:

.. list-table::
   :widths: 35 65 
   :header-rows: 1

   * - Event Type
     - Description

   * - :php:`ServerChangedEvent <mongodb-driver-monitoring-serverchangedevent>`
     - Created when the server description changes, such as the server's type
       changing from secondary to primary.

   * - :php:`ServerOpeningEvent <mongodb-driver-monitoring-serveropeningevent>`

>      - Created when the server description is instantiated with its

Reviewer 2: For PHP, I think this can say "when a server connection is established". I'm asking for feedback on this in the drivers channel, but I don't think the spec's language about instantiating a server description is relevant to users.

For other drivers, I expect the public-facing docs would be "when the driver establishes a monitoring connection to the server".

Depending on the feedback I receive, perhaps we may want to say "immediately before the driver establishes a connection to the server" (or "monitoring connection to the server" for other drivers).

Reviewer 2: @rustagir: Feedback came back. I think we're good to simply say "when a server connection is established".

       defaults.

   * - :php:`ServerClosedEvent <mongodb-driver-monitoring-serverclosedevent>`

>      - Created when the server is closed.

Reviewer 2: This can say "when a server connection is closed". For other drivers, this only pertains to the monitoring thread's connections, but monitoring and app connections are one and the same in PHP since our SDAM implementation is single-threaded.

   * - :php:`TopologyChangedEvent <mongodb-driver-monitoring-topologychangedevent>`
     - Created when the topology description changes, such when there is

>        an election of a new primary or disconnection of a ``mongos`` proxy.

Author: Ah, that must be a copy paste artifact.

   * - :php:`TopologyOpeningEvent <mongodb-driver-monitoring-topologyopeningevent>`
     - Created when the topology description is initialized.

   * - :php:`TopologyClosedEvent <mongodb-driver-monitoring-topologyclosedevent>`

>      - Created when the topology is closed.

Author: thanks for providing this background!

   * - :php:`ServerHeartbeatStartedEvent <mongodb-driver-monitoring-serverheartbeatstartedevent>`
     - Created when the server monitor sends a ``hello`` call to the server.
       This action is called a heartbeat.

   * - :php:`ServerHeartbeatSucceededEvent <mongodb-driver-monitoring-serverheartbeatsucceededevent>`
     - Created when the heartbeat succeeds.

   * - :php:`ServerHeartbeatFailedEvent <mongodb-driver-monitoring-serverheartbeatfailedevent>`
     - Created when the heartbeat fails.

You can find a list of the monitoring subscriber classes and event
methods in the :php:`Monitoring classes and subscriber functions
<mongodb.monitoring>` section of the PHP manual.

Remove a Subscriber
-------------------

Later in your application, you might not want to subscribe to
SDAM events. To unregister a subscriber from your client, use the
``MongoDB\Client::removeSubscriber()`` method. If you attempt to remove

> a nonexistent subscriber, the method doesn't perform any action.

Reviewer 2: On a related note, PHPC (extension) also doesn't allow you to register the same subscriber twice. Attempting to do so would also be a no-op. I don't think we need to mention that, though.
